### PR TITLE
More DQMC measurements

### PR DESCRIPTION
**Pull request by ffreyer**
This commit adds:

* `greens(mc, slice)` to compute the greens function for an arbitrary time
* `ChargeDensityCorrelationMeasurement` to measure n_i n_j
* `MagnetizationMeasurement`to measure s_x, s_y and s_z
* `SpinDensityCorrelationMeasurement` to measure `s_{x, i} s_{x, j}`, `s_{y_i} s_{y, j}` and `s_{z, i} s_{z, j}`
* `PairingCorrelationMeasurement` measuring `ΔΔ^† = c_i c_i c_j^† c_j^†`

Currently tested:
* `PairingCorrelationMeasurement` and occupation (from averaged Greens function) for triangular Hubbard matches https://journals.aps.org/prb/abstract/10.1103/PhysRevB.48.3976

TODO:
- Add tests comparing to ED results
  - [x] `ChargeDensityCorrelationMeasurement`
  - [x] `MagnetizationMeasurement` (mostly useless due to symmetries in G)
  - [x] `SpinDensityCorrelation`
  - [x] `PairingCorrelationMeasurement`
- ~~Add tests comparing to literature~~ Comparing to ED should be enough
  -  `ChargeDensityCorrelationMeasurement`
  -  `MagnetizationMeasurement`
  -  `SpinDensityCorrelation`
  -  `PairingCorrelationMeasurement` (but currently way to slow)
- [x] Test `greens(mc, slice)`
-  ~~Add unequal-time greens function `greens(mc, slice1, slice2)`~~

### CODE REVIEW

**Comment by ffreyer** [comment]
Some Notes based on discussions with Carstens and issues I found

* `greens(mc, slice1, slice2)`, as it is currently implemented, is unstable (Why?)
* `greens(mc, slice1, slice2)` might be wrong in general, I have to check. (According to dos Santos Intro to DQMC we get G or 1-G depending on which slice is later in time.)
* `greens(mc, slice)` still needs to be checked, but that should be easy once I fix the functions in `test/testfunctions.jl`.
* DQMC Measurements often require a lot of memory. So rather than setting a high default capacity for `LightObservable`s \ `LogBinner`s, they should inherit a reasonable capacity from the number of sweeps in dqmc.
* Carsten mentioned that CDW also come in s-wave, d-wave, etc variants. This should not be a problem if these characteristics can be derived from the lattice geometry. (like done [here](https://journals.aps.org/prb/pdf/10.1103/PhysRevB.72.134513) for the pairing correlation on a triangular lattice)
* Since the definitions of of observables are (at least somewhat) arbitrary it would be good to make them easy to change. I.e. have a kernel function in the to fill the CDW correlation matrix rather than an explicit formula.
* We can measure `n_up - n_down` (or `n_up + n_down -1` depending on the sign of U) using the auxiliary field.
* The s-wave / uniform_fourier part of the pairing correlation measurement may be different between models, so it should not be part of the measurement. It should also be the same if the calculated after the simulation. (check this)

**Comment by MasonProtter** [comment]
Any literature suggestions for implementing comparisons? It's hard to find literature that studies small enough systems that the CI can do it.

**Comment by carstenbauer** [comment]
For small or non-interacting systems we should compare to ED or exact results, if possible.

**Comment by ffreyer** [comment]
In #50 Carsten noticed that ED was using the wrong sign for U. With the merge (b3ca144) this is now fixed. 
As a result all measurements now fit ED up to an absolute error of delta_tau^2 or a relative error of 1%. (The relative error only affects two charge density correlations)

**Comment by ffreyer** [comment]
I changed `measurements` and `observables` to return only measurements that are recorded after the thermalization stage. 
I also added `mean(measurement)`, `var(measurement`, `std_error(measurement` and `tau(measurement)`. These methods look for observables in a measurement and then call `MonteCarloObservable.mean()` etc. on each observable.

**Comment by ffreyer** [comment]
I started porting the unequal time Greens function from @crstnbr private repository. TODO
* check whether the implementation is correct
* add tests
* figure out how to integrate it (The unequal time Greens function requires its own propagator stacks. Where should these be kept?)

**Comment by ffreyer** [comment]
I decided to deal with unequal time Greens functions in another pr

**Comment by ffreyer** [comment]
tl;dr for the last 20 commits:
The summation in `measure!` was wrong for most observables, because it disregarded the lattice. Masks fix this, with `RawMask` resulting in no summation at runtime and `DistanceMask` resulting in the correct summation. (Tested with and without basis for a bunch of 2D and 3D LatticePhysics lattices)

**Comment by ffreyer** [comment]
TODO:
- [x] try to compress `PairingCorrelationMeasurement`
* ~#76~ (maybe later in another pr)

**Comment by ffreyer** [comment]
`SystemError: realpath: No such file or directory`
... just one mac. Cool.

I think I'm at a point where I'm happy to merge this. Might do some more small tweaks though.
